core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,374 @@
+    *
+    * If appending the new state would introduce a duplicate state name, then a
+    * cycle has been found. Specifically, if pushing the newest state (N) onto
+    * the path would result in the following pattern:
+    *
+    *    [ ... , N , ... , N ]
+    *
+    * then we update this tracker's latest cycle discovery. The cycle is defined
+    * as the current path of visited states with all states prior to the first
+    * occurrence of N discarded:
+    *
+    *    [ ... // N , ... , N ]  =>  [ N , ... , N ]
+    *
+    * If this *exact* cycle has been discovered before (NOTE: rotations of a
+    * cycle are considered to be different from each other),then we increase the
+    * tracked number of cycle occurrences.
+    * Otherwise, the tracker is completely reset to 1.
+    *
+    * Finally, the path of visited states is reset to a value of [ N ], in hopes
+    * that the exact same cycle is discovered enough times to be considered an
+    * issue.
+    */
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName() === destStateName) {
+      return;
+    }
+    if (this.pathOfVisitedStates.includes(destStateName)) {
+      const cycleOfVisitedStates = (
+        this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
+      if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
+        this.numLoops += 1;
+      } else {
+        this.cycleOfVisitedStates = cycleOfVisitedStates;
+        this.numLoops = 1;
+      }
+       this.pathOfVisitedStates.length = 0;
+    }
+    this.pathOfVisitedStates.push(destStateName);
+  }
+
+  generateIssueCustomizationArgs(): CyclicStateTransitionsCustomizationArgs {
+    return {
+      state_names: {value: this.cycleOfVisitedStates}
+    };
+  }
+
+  private makeCycle(collisionIndex: number): string[] {
+    const collision = this.pathOfVisitedStates[collisionIndex];
+    const cycleWithNoCollision =
+    this.pathOfVisitedStates.slice(collisionIndex);
+    return [...cycleWithNoCollision, collision];
+  }
+
+  private currStateName(): string {
+    return this.pathOfVisitedStates[this.pathOfVisitedStates.length - 1];
+  }
+}
+
+class EarlyQuitTracker {
+  private stateName: string = ' ';
+  private expDurationInSecs: number | null = null;
+
+  foundAnIssue(): boolean {
+    return (
+      this.expDurationInSecs !== null &&
+      this.expDurationInSecs < ServicesConstants.EARLY_QUIT_THRESHOLD_IN_SECS);
+  }
+
+   recordExplorationQuit(
+       stateName: string,
+       expDurationInSecs: number | null
+   ): void {
+    if (this.stateName !== null && this.expDurationInSecs !== null) {
+      this.stateName = stateName;
+      this.expDurationInSecs = expDurationInSecs;
+    }
+  }
+
+
+  generateIssueCustomizationArgs(): EarlyQuitCustomizationArgs {
+    if (this.expDurationInSecs !== null) {
+      return {
+        state_name!: {value: this.stateName},
+        time_spent_in_exp_in_msecs!: {value: this.expDurationInSecs * 1000},

```

Comment:
Why you add the `!` here?

Reply:
I didn't added null here ,it was here all the time  and if i remove it now it will cause error.